File: app/src/lib/api/ml.ts
import { apiGet, apiPost } from "./client";

export type TrainResult = {
  num_samples: number;
  num_categories: number;
  cv_accuracy: number | null;
};

export type Suggestion = {
  transaction_id: number;
  category_id: number;
  category_name: string;
  confidence: number;
  payee: string | null;
  purpose: string | null;
  amount: string;
  booking_date: string;
  could_become_rule: boolean;
};

export function trainModel(): Promise<TrainResult> {
  return apiPost<TrainResult>("/api/ml/train", {});
}

export function getSuggestions(): Promise<Suggestion[]> {
  return apiGet<Suggestion[]>("/api/ml/suggest");
}

File: api/my_private_finances/schemas/ml.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from sqlmodel import SQLModel


class TrainResult(SQLModel):
    num_samples: int
    num_categories: int


class Suggestion(SQLModel):
    transaction_id: int
    category_id: int
    category_name: str
    confidence: float
    payee: str | None
    purpose: str | None
    amount: Decimal
    booking_date: date
